ibm-openbmc/bmcweb
# 1110: Pickup heartbeat logic from upstream

This implements  2 stories 
- https://jsw.ibm.com/browse/PFEBMC-2690 // Persist the heartbeat setting at BMC
- https://jsw.ibm.com/browse/PFEBMC-2688 //  Implement SendHeartbeat and HeartbeatIntervalMinutes properties

The following commits are are picked from upstream ;

- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/75791  // Implement Subscription Heartbeat Logic
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/75516  // Add Persistent Heartbeat subscription properties
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/74180  // event service: dbus log: enable event subscription
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/76023  // Break out Filesystem events into classes
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/76121  // Break out SendEventToSubs
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/76020  // Break out long lambda in filesystem event
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/76021  // Break out Dbus events into class
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/76019  // Break out dbus telemetry watcher
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/76018  // Break out filesystem watcher into separate file
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/75891  // Refactor: move registries functions to their file
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/75890  // Refactor: extract src/event_log.cpp
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/75235  // Refactor: break up event_service_manager.hpp
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/75433  // Make UserSubscription as shared_ptr in Subscription
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/75459  // Remove SubscriptionId get and set functions
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/74731  // event logs: introduce struct DbusEventLogEntry
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/74179  // event service manager: more meaningful debug logs
- https://gerrit.openbmc.org/c/openbmc/bmcweb/+/71181  // clang-format: re-format for clang-18

In addition, 1110 adds a local change after `https://gerrit.openbmc.org/c/openbmc/bmcweb/+/71181 `.

- af329dc3a Fix clang-format mishandling of assembly.hpp
It is simply adding  `\\` at the end of a line of `assembly.hpp`  so that `clang-format-18` can produce a better readable format.

assembly.hpp:
```
       sdbusplus::asio::getProperty<associationList>(*crow::connections::systemBus, assemblyParentServ,
-->
       sdbusplus::asio::getProperty<associationList>( //
            *crow::connections::systemBus, assemblyParentServ,
```

Tested:

- Redfish Service Validator passes as it was working before (i.e. Cable will have known issues)
- Create subscription - either POST or PATCH, and check the content of GET

For example,
```
curl -k -X PATCH [https://${bmc}/redfish/v1/EventService/Subscriptions/${SUBID](https://${bmc}/redfish/v1/EventService/Subscriptions/${SUBID)} \
    -H "Content-Type: application/json" \
     -d '{"SendHeartbeat":true, "HeartbeatIntervalMinutes":10}'
```

- Use Redfish Event Listener and monitor the heartbeat messages periodically (as it is defined as `HeartbeatIntervalMinutes`)

```
response_type: POST
headers: {'Host': '9.3.62.209', 'Content-Length': '230'}

bodydata: {
  "@odata.type": "#Event.v1_4_0.Event",
  "Events": [
    {
      "@odata.type": "#Message.v1_1_1.Message",
      "EventId": 1,
      "EventTimestamp": "2024-11-13T17:20:14+00:00",
      "MemberId": "0",
      "Message": "Redfish service is functional.",
      "MessageArgs": [],
      "MessageId": "HeartbeatEvent.1.0.1.RedfishServiceFunctional",
      "MessageSeverity": "OK",
      "OriginOfCondition": "",
      "Resolution": "None."
    }
  ],
  "Id": 1,
  "Name": "Event Log"
}
```

- PATCH to stop SendHeartbeat, and see Redfish Event Listener not receiving the heartbeat
```
curl -k -X PATCH [https://${bmc}/redfish/v1/EventService/Subscriptions/${SUBID](https://${bmc}/redfish/v1/EventService/Subscriptions/${SUBID)} \
    -H "Content-Type: application/json" \
     -d '{"SendHeartbeat":false}'
```

- Stop Event Listener and wait for more than ( `DeliveryRetryAttempts` * `HeartbeatIntervalMinutes` ), check whether the subscription is deleted.
